keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1330 @@
+  phase of the maintenance.
+
+#### Supported Stage Transitions
+
+The following transitions should be validated by the API server.
+
+- Idle -> _Deletion_
+  - Planning a maintenance in the future and canceling/deleting it without any consequence.
+- (Idle) -> Cordon -> (Complete) -> _Deletion_.
+  - Make a set of nodes unschedulable and then schedulable again.
+  - The complete stage will always be run even without specifying it.
+- (Idle) -> (Cordon) -> Drain -> (Complete) -> _Deletion_.
+  - Make a set of nodes unschedulable, drain them, and then make them schedulable again.
+  - Cordon and Complete stages will always be run, even without specifying them.
+- (Idle) -> Complete -> _Deletion_.
+  -  Make a set of nodes schedulable.
+
+### Deployment and ReplicaSet Controllers
+
+The replica set controller will watch its pods and count the number of pods it observes with a
+`EvacuationRequested` condition. It will then store this count in `.status.ReplicasToEvacuate`
+
+The deployment controller will watch its ReplicaSets and react when it observes positive number of
+pods in `.status.ReplicasToEvacuate`. If the owning object of the targeted pods is a Deployment with
+a positive `.spec.strategy.rollingUpdate.maxSurge` value, the controller will create surge pods by
+scaling up the ReplicaSet. The new pods will not be scheduled on the maintained node because the
+`.spec.unschedulable` field would be set to true on that node. As soon as the surge pods become
+available, the deployment controller will scale down the ReplicaSet. The replica set controller
+will then in turn delete the pods with the `EvacuationRequested` condition.
+
+For completeness, the deployment controller will also track the total number of targeted pods of
+all its ReplicaSets under its `.status.ReplicasToEvacuate`.
+
+If the node maintenance prematurely ends before the surge process has a chance to complete, the
+deployment controller will scale down the ReplicaSet which will then remove the extra pods that
+were created during the surge.
+
+```golang
+type ReplicaSetStatus struct {
+    ...
+    ReplicasToEvacuate int32
+    ...
+}
+```
+
+```golang
+type DeploymentStatus struct {
+    ...
+    ReplicasToEvacuate int32
+    ...
+}
+```
+
+To support providing a response to the drain process that the evacuation has begun. The deployment
+controller will annotate all replica sets that support the evacuation with the annotation
+`deployment.kubernetes.io/evacuation-ready`. For now this will apply to replica sets of deployments
+with `.spec.strategy.rollingUpdate.maxSurge`. When this annotation is present, the replication
+controller will respond to all evacuation requests by setting the `EvacuationInitiated` condition
+to all of its pods with the `EvacuationRequested` condition.
+
+
+### Node Drain Process: Before and After
+
+The following diagrams describe how the node drain process will change in respective to each
+component.
+
+Current state of node drain:
```

Comment:
The diagram above is over simplified.
A drain consumes the eviction api for all the pods of the nodes, and not the Delete. This is a huge difference, because the eviction API creates the an eviction subresource. This creation lead to an admission-webhook call that allows platform admins to do a LOT that is going to be skipped with the solution. In our company we are doing multiple checks at eviction admission that will not be possible anymore with the solution. 
That would be a problem, for any user that is gating the pod eviction at this level to:
- validate that there is no global context that prevent that eviction. (During incident it could be convenient to block eviction by CA (repacking) or by other automation), (During operation freeze it could be interesting to block any user eviction request)...
- perform last synchronous check: if you application is distributed over multiple clusters for resiliency purpose, you don't want to create disruption/eviction in all the clusters at the same time.

To summarise, the loss of the admission-webhook on evictions could have a huge impact on operation by removing a convenient point of control. The proposed solution should come with similar capability. If not we are going to solve a problem but we are going to create another one.

Reply:
Thanks for the review, you raise a good point here.

I actually think people might want a node maintenance controller to attempt graceful (eviction subresource) draining, and then optionally switch to forcible Pod removal.

For example, if the data center is scheduled to get powered down Tuesday, blocking the API-driver eviction with a webhook response or ValidatingAdmissionPolicy isn't going to keep the electricity supply going.

Reply:
Thank you for this important feedback. The admission webhooks intercepting the evictions are not supported in Kubernetes today. This introduces a lot of undefined behaviour that cannot be reasoned about by other components using the eviction API.

> To summarise, the loss of the admission-webhook on evictions could have a huge impact on operation by removing a convenient point of control. The proposed solution should come with similar capability. If not we are going to solve a problem but we are going to create another one.

This KEP and the Evacuation KEP https://github.com/kubernetes/enhancements/pull/4565 are trying to solve this problem. That is, you can subscribe to the evacuations and become the evacuator who then decides how the evacuation should proceed and wait for the global context and do before deletion checks.

The Evacuation object itself can be used as a lock to synchronize the evacuator ownership. It does not discuss the evacuator priority though, but I will try to outline it there. 

Even though the webhook approach is convenient, it is very opaque and hard to debug, so we try to go with a transparent approach.

@dbenque 
- Could you please open an issue with this feature requres in kubernetes/kubernetes so we can reference it in the KEP? 
- What mechanism do you think would be the best for handing over the ownership between the different evacuators?

Reply:
@atiratree can you please point me at the documentation or code that justifies the following:
> The admission webhooks intercepting the evictions are not supported in Kubernetes today

I am a bit concerned by this statement because I have been using such an admission on eviction for more than 2y now and it works fine and it gives full satisfaction.

I will have a look to KEP https://github.com/kubernetes/enhancements/pull/4565 thanks for the link.

> it is very opaque and hard to debug, so we try to go with a transparent approach.

Not sure what you mean by "opaque", admission is a well known pattern in kubernetes. Sure it can be a double edge sword, and this is why as platform owner we try not to let users define/play with it... but for a platform owner it is a very powerful and convenient tool.
If we find a better way that offer the same capability (level of control) then that should be fine.

Reply:
(I'm not sure about unsupported, but both the v1 Pod API and admission control through validating webhooks are generally available features)

Reply:
> I am a bit concerned by this statement because I have been using such an admission on eviction for more than 2y now and it works fine and it gives full satisfaction.

Admission webhooks are supported, but not everything what you do with them is. For example, if you fail all pod creations, you cannot be dissatisfied that the ReplicaSet does not create the pods and thus failing the conformance. The same goes for eviction and all the places that depend on it.

> Not sure what you mean by "opaque", admission is a well known pattern in kubernetes. Sure it can be a double edge sword, and this is why as platform owner we try not to let users define/play with it... but for a platform owner it is a very powerful and convenient tool.

It is opaque in the sense that it is not discoverable and observable by other actors who have not deployed or developed the admission plugins.

@dbenque I have added a support for multiple evacuators to the Evacuation KEP https://github.com/kubernetes/enhancements/pull/4565#issuecomment-2045722703 . It also tackles the issues of discoverability and observability. Please let me know if it solves your use case.